### ISSUE TITLE
Types of fields could change across await boundary

### ISSUE THREAD

**Issue by michalburger1** [microsoft/TypeScript#35231]
**TypeScript Version:**  3.7.2

**Search Terms:** await this field modified

**Code**

```ts
async function sleep(time: number) {
    return new Promise((resolve) => setTimeout(resolve, time))
}

function doNumberThings(num: number) {
    console.log(num)
    console.log(num ** 2)
}

class Foo {
    public bar: number | string = 123

    public async checkBar() {
        if (typeof this.bar === 'number') {
            await sleep(1000)
            doNumberThings(this.bar)
        }
    }
}

const foo = new Foo()
setTimeout(() => foo.bar = 'oops', 500)
foo.checkBar()
```

**Expected behavior:** Type checker complains that this.bar could be a string

**Actual behavior:** Type checker does not complain, this.bar is string at runtime

**Playground Link:** http://www.typescriptlang.org/play/?ssl=1&ssc=1&pln=24&pc=1#code/IYZwngdgxgBAZgV2gFwJYHsIxAGwKZ4AOAFGgLZ4BcMECZARngE4CUMA3gLABQMfMTPMgRMsEPAHcYABSboyqEHmLFBIdDgBueNgF4AfNiEAVVBXQJkqvOq14ANDHI6WPAL48eiFBiwATdAA5OkYmYwALVAgAcxBiWjJqBNC2Ll5+KExbPAA6HHRo+LpXdL5MiGy8gqKyGAAqOpgAJhKPbh4oHFAQGAAxdHQOHn4YQgR6HFRYemAmJJDmGAAfbGQmKOiYXRgARiaAZk9S0fHJ2FBIWChwvCgAawAhWeJU4ZH+VDgYUjBCPHQvshIiAcjMmFtdNsAOTJZhQ17Hd78YASYCoZDYfBEYg7AAM+JKSKRAWCDGYEQ2cSBilBs0JRLaIzabQ6WQxcAGWxokj6AxePCUyFM5ksKj0hg56Fp4OhA0IIChjgArASvAMctdbo9niUgA

**Related Issues:**

**Comment by MartinJohns** [comment]
Duplicate of #9998.
